pkg/sparrow/run.go

```diff
@@ -137,118 +141,93 @@ func (s *Sparrow) Run(ctx context.Context) error {
-
-		c := s.updateCheckTargets(checkCfg)
-		if existingCheck, ok := s.checks[name]; ok {
-			// Check already registered, reset config
-			err := existingCheck.SetConfig(ctx, c)
+func (s *Sparrow) ReconcileChecks(ctx context.Context, cfg runtime.Config) {
+	// generate checks from configuration
+	s.enrichTargets(cfg)
+	fChecks, err := factory.NewChecksFromConfig(cfg)
+	if err != nil {
+		logger.FromContext(ctx).ErrorContext(ctx, "Failed to create checks from config", "error", err)
+		return
+	}
+
+	// if checks are empty, register all checks
+	if len(s.checks) == 0 {
+		for _, c := range fChecks {
+			err = s.registerCheck(ctx, c)
 			if err != nil {
-				log.ErrorContext(ctx, "Failed to reset config for check, check will run with last applies config", "error", err)
+				logger.FromContext(ctx).ErrorContext(ctx, "Failed to register check", "error", err)
 			}
-			continue
 		}
-
-		// Check is a new Check and needs to be registered
-		s.registerCheck(ctx, name, c)
+		return
 	}
 
-	for existingCheckName, existingCheck := range s.checks {
-		if _, ok := s.cfg.Checks[existingCheckName]; ok {
-			// Check is known check
-			continue
+	// unregister checks that are not in the new config
+	for name, check := range s.checks {
+		if !cfg.HasCheck(name) {
+			s.unregisterCheck(ctx, check)
 		}
-
-		// Check has been removed from config
-		s.unregisterCheck(ctx, existingCheckName, existingCheck)
-	}
-}
-
-// updateCheckTargets updates the targets of a check with the
-// global targets. The targets are merged per default, if found in the
-// passed config.
-func (s *Sparrow) updateCheckTargets(cfg any) any {
-	if cfg == nil {
-		return nil
-	}
-
-	// check if map with targets
-	checkCfg, ok := cfg.(map[string]any)
-	if !ok {
-		return checkCfg
-	}
-	if _, ok = checkCfg["targets"]; !ok {
-		return checkCfg
 	}
 
-	// Check if targets is a slice
-	actuali, ok := checkCfg["targets"].([]any)
-	if !ok {
-		return checkCfg
-	}
-	if len(actuali) == 0 {
-		return checkCfg
-	}
+	// register / update checks that are in the new config
+	for _, c := range fChecks {
+		if _, ok := s.checks[c.Name()]; !ok {
+			err = s.registerCheck(ctx, c)
+			if err != nil {
+				logger.FromContext(ctx).ErrorContext(ctx, "Failed to register check", "error", err)
+			}
+			continue
+		}
 
-	// convert to string slice
-	var actual []string
-	for _, v := range actuali {
-		if _, ok := v.(string); !ok {
-			return checkCfg
+		// existing config
+		err = s.checks[c.Name()].SetConfig(c.GetConfig())
+		if err != nil {
+			logger.FromContext(ctx).ErrorContext(ctx, "Failed to set config for check", "error", err)
 		}
-		actual = append(actual, v.(string))
 	}
-	var urls []string
+}
 
-	if s.tarMan == nil {
-		return checkCfg
+// enrichTargets updates the targets of the sparrow's checks with the
+// global targets. Per default, the two target lists are merged.
+func (s *Sparrow) enrichTargets(cfg runtime.Config) runtime.Config {
+	if cfg.Empty() || s.tarMan == nil {
```

Comment:
Since there currently is no validation (outside the YAML syntax one) when grabbing a new `runtime.Config`, the following config may also be pushed to the channel:

```
incoming := runtime.Config{}

// Parsed from an empty YAML file, which is a legit option
```

We should either stop something from this ever happening at loader-time, or just write a simple guard here, which makes more sense to me.

The point about reworking the function is valid.